tgui/packages/tgui/interfaces/SpawnPanel/CreateObject.tsx

```diff
@@ -0,0 +1,356 @@
+    search_by: string;
+  };
+}
+
+export function CreateObject(props: CreateObjectProps) {
+  const { act, data } = useBackend<SpawnPanelData>();
+  const { objList } = props;
+
+  const [tooltipIcon, setTooltipIcon] = useState(false);
+  const [selectedObj, setSelectedObj] = useState<string | null>(null);
+
+  const [searchText, setSearchText] = useState('');
+  const [searchBy, setSearchBy] = useState(false);
+  const [sortBy, setSortBy] = useState(listTypes.Objects);
+  const [hideMapping, setHideMapping] = useState(false);
+  const [showIcons, setshowIcons] = useState(false);
+  const [showPreview, setshowPreview] = useState(false);
+  const currentType =
+    Object.entries(listTypes).find(([_, value]) => value === sortBy)?.[0] ||
+    'Objects';
+
+  const currentList = objList?.[currentType] || {};
+
+  useEffect(() => {
+    if (data.selected_object) {
+      setSelectedObj(data.selected_object);
+    }
+  }, [data.selected_object]);
+
+  useEffect(() => {
+    if (data.copied_type) {
+      setSelectedObj(data.copied_type);
+      setSearchText(data.copied_type);
+      if (objList.Turfs[data.copied_type]) {
+        setSortBy(listTypes.Turfs);
+      } else if (objList.Mobs[data.copied_type]) {
+        setSortBy(listTypes.Mobs);
+      } else {
+        setSortBy(listTypes.Objects);
+      }
+      setSearchBy(true);
+    }
+  }, [data.copied_type, objList]);
+
+  useEffect(() => {
+    const loadStoredValues = async () => {
+      const storedSearchText = await storage.get('spawnpanel-searchText');
+      const storedSearchBy = await storage.get('spawnpanel-searchBy');
+      const storedSortBy = await storage.get('spawnpanel-sortBy');
+      const storedHideMapping = await storage.get('spawnpanel-hideMapping');
+      const storedShowIcons = await storage.get('spawnpanel-showIcons');
+      const storedShowPreview = await storage.get('spawnpanel-showPreview');
+
+      if (storedSearchText) setSearchText(storedSearchText);
+      if (storedSearchBy !== undefined) setSearchBy(storedSearchBy);
+      if (storedSortBy) setSortBy(storedSortBy);
+      if (storedHideMapping !== undefined) setHideMapping(storedHideMapping);
+      if (storedShowIcons !== undefined) setshowIcons(storedShowIcons);
+      if (storedShowPreview !== undefined) setshowPreview(storedShowPreview);
+    };
+
+    loadStoredValues();
+  }, []);
+
+  useEffect(() => {
+    storage.set('spawnpanel-searchText', searchText);
+  }, [searchText]);
+
+  useEffect(() => {
+    storage.set('spawnpanel-searchBy', searchBy);
+  }, [searchBy]);
+
+  useEffect(() => {
+    storage.set('spawnpanel-sortBy', sortBy);
+  }, [sortBy]);
+
+  useEffect(() => {
+    storage.set('spawnpanel-hideMapping', hideMapping);
+  }, [hideMapping]);
+
+  useEffect(() => {
+    storage.set('spawnpanel-showIcons', showIcons);
+  }, [showIcons]);
+
+  useEffect(() => {
+    storage.set('spawnpanel-showPreview', showPreview);
+  }, [showPreview]);
+
+  const sendPreferences = (settings) => {
```

Comment:
make this a typed function declaration if you're a bad enough dude. is settings Record<string, unknown>?